### ISSUE TITLE
[Tabix] Unexpected chunk virtual offsets with more than >1 chunk bins

### ISSUE THREAD

**Issue by matty234** [samtools/htslib#1864]
I'm trying to understand the Tabix file format. I've been working through both the HTSLib documentation and the Tabix spec and I'm struggling to understand the virtual indexes in a specific scenario.

I'm parsing the data fine up until the number of chunks in a bin > 1. At which point, I always find that the second chunk has a tiny start virtual offset follow by a `0` as the end virtual offset.

As an example, I've annotated two bins from a truncated tabix file:

[**Bin 1** (with one chunk):
](https://hexplore.dev/shared/9fe77b41-7a30-4961-8cc9-7f356cca51bf#comment-720-723)

> 
> Chunk 1 begin virtual offset: [542993156763](https://hexplore.dev/shared/9fe77b41-7a30-4961-8cc9-7f356cca51bf#comment-728-735)
> Chunk 1 end virtual offset: [542993157071](https://hexplore.dev/shared/9fe77b41-7a30-4961-8cc9-7f356cca51bf#comment-736-743)

which all makes sense to me!

However, with **[Bin 2](https://hexplore.dev/shared/9fe77b41-7a30-4961-8cc9-7f356cca51bf#comment-1152-1155)** (which has [two chunks](https://hexplore.dev/shared/9fe77b41-7a30-4961-8cc9-7f356cca51bf#comment-1156-1159)):

> Chunk 1 begin virtual offset: [150274048](https://hexplore.dev/shared/9fe77b41-7a30-4961-8cc9-7f356cca51bf#comment-1160-1167)
> Chunk 1 end virtual offset: [2078376086007](https://hexplore.dev/shared/9fe77b41-7a30-4961-8cc9-7f356cca51bf#comment-1168-1175)
> Chunk 2 begin virtual offset: [362756](https://hexplore.dev/shared/9fe77b41-7a30-4961-8cc9-7f356cca51bf#comment-1176-1183)
> Chunk 3 end virtual offset: [0](https://hexplore.dev/shared/9fe77b41-7a30-4961-8cc9-7f356cca51bf#comment-1184-1191)

which doesn't make sense from my understanding of the format.

This is likely just an issue with my comprehension of the specification; apologies if so.

**Comment by matty234** [comment]
I made this shareable hex browser specifically for this issue so apologies if it's a little rough around the edges 😄

**Comment by daviesrob** [comment]
If you haven't already done so, you could try building htslib with:
```
make CPPFLAGS='-DDEBUG_INDEX'
```

That will cause the function [`idx_dump()`](https://github.com/samtools/htslib/blob/c705bec2af9caedacc6ba80dfba7fe625eb5ba5c/hts.c#L2264) to be built and called whenever you create or read an index, and dump the contents to `stderr`.  Hopefully that will help you understand what's going on.

**Comment by daviesrob** [comment]
Yes, arguably the meta bins are only [documented](https://samtools.github.io/hts-specs/SAMv1.pdf#subsection.5.2) for the SAM-specific BAI and CSI index formats, and not TBI as you had, so possibly shouldn't be there.  They are mostly harmless though as the bin number used should never be reachable in a normal index query.
